react components (#36)
Ada-Developers-Academy/core-unit-3: react-components/intro-to-react-js.md

# Intro to React JS

## Learning Goals

> - What is React?
> - Why do we learn React?
> - Who is using React?
> - Intro to the Virtual DOM
> - Intro to Components

Reviewer: these don't sound very goal-like at the moment. suggest a rewording like the following (feel free to tweak!).

also, I didn't see anything in here about the virtual DOM, so maybe remove that from here? or if this is meant to be more over-arching learning goals for React as a whole, then maybe include some clarification about when that goal will be met.

```suggestion
- Define what React is
- Understand why we learn React
- Discuss who is using React
- Acquire introductory understanding of React Components
```

## Introduction

> Consider a social media website, such as [Reddit.com](https://www.reddit.com/r/reactjs/). Every given web page needs to do soooo many things! To name a few, each web page needs to:

Reviewer: restrict pages under discussion to those on Reddit.
"things" doesn't feel like a concrete-enough noun to omit it from the "To name a few" phrase, but adding it in sounds redundant. I avoided the issue by switching over to "for example".
```suggestion
Consider a social media website, such as [Reddit.com](https://www.reddit.com/r/reactjs/). Every web page on that site needs to do soooo many things! For example, each web page needs to:
```

> - List posts, the number of upvotes, and the comments

Reviewer: at least in the screen shot, upvotes and comments both seem to be numbers, and both are related to a post, so maybe the following phrasing to reflect that?

if this seems like an unnatural interpretation to a more native redittor, feel free to ignore!

```suggestion
- List posts, including the numbers of upvotes and comments
```

- Display upvote and downvote buttons
- Handle upvote and downvote buttons
- List related communities
- Handle the buttons to "join" those communities
- Display and handle the UI to create a new post

> ![](../assets/react-components_intro-to-react-js_reddit.png)

Reviewer: adds alt text and caption
```suggestion
![The UI layout of the /r/ReactJS subreddit, made up of many distinct regions with specific responsibilities.](../assets/react-components_intro-to-react-js_reddit.png)  
_Fig. The /r/ReactJS subreddit. The self-styled "Front Page of React"_
```

> Imagine creating all of this functionality using HTML and vanilla JavaScript. It wouldn't be impossible, but our resulting code is probably:

Reviewer: we're imagining, so use a more hypothetical phrasing
```suggestion
Imagine creating all of this functionality using HTML and vanilla JavaScript. It wouldn't be impossible, but our resulting code would probably be:
```

- Huge! There's a lot of code in that logic to write!
- Unorganized. Is there one file that holds all the JS? Are there multiple files? How are they organized?

> - Full of repetition. We need to create event handlers all the time. Is there a way to simplify this repetition?

Reviewer: I mean, not _really_ "all the time". but a lot. possible rephrasing?
```suggestion
- Full of repetition. We need to create many, many event handlers. Is there a way to simplify this repetition?
```

> Great software is maintainable, functional, and performant. How can we strive towards those values in our front-end code?

Reviewer: I've run into this a few times lately with performant. It doesn't actually appear in the dictionary (!), and we haven't ever officially defined it in our curriculum as required jargon. It's easily avoided, so maybe rephrase?
```suggestion
Great software is maintainable, functional, and performs well. How can we strive towards those values in our front-end code?
```

Author: ooOOOOoh thank you! I think I end up using the word "performant" in another PR, so sorry!!! I'm taking notes.

Reviewer: I don't think I even realized it wasn't a "standard" word until the spell checker yelled and me and I went looking for more info. I definitely use performant in conversation myself (and GH here isn't complaining about it), but it's definitely tech jargon.

Maybe I have a bit too prescriptive a view of word usage, but now that I know it's "non-standard" I'd be lying if I didn't say it makes me cringe a little. 🤣

## Vocabulary and Synonyms

| Vocab               | Definition                                                                                                                                                                                               | How to Use in a Sentence                                                                                                                                                                                               |
| ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | User Interface (UI) | The elements that a user uses in order to interact with a machine. UI elements are usually visual, but UI also describes _how_ a user interacts with a webapp, such as buttons, text fields, or sliders. | "This UI was easy to use and enjoyable, because the text was readable, and the buttons were intuitive," "This UI was difficult to use, because I didn't understand the navigation menu, and the data was unorganized." |

Reviewer: rephrase to avoid user uses
```suggestion
| User Interface (UI) | The elements with which a user interacts in order to make a machine do something. UI elements are usually visual, but UI also describes _how_ a user interacts with a webapp, such as buttons, text fields, or sliders. | "This UI was easy to use and enjoyable, because the text was readable, and the buttons were intuitive," "This UI was difficult to use, because I didn't understand the navigation menu, and the data was unorganized." |
```

| Components (React)  | Independent, reusable pieces of UI.                                                                                                                                                                      | "My recipe webapp has a `Recipe` component, which renders a recipe," "I created a `RedButton` component to reuse. All `RedButton` components are red and animate on click."                                            |

## Create Complex UIs with React Components

> React is an [open-source](https://github.com/facebook/react) JavaScript library for building user interfaces.

Reviewer: I feel like the link placement is a bit odd... I would have assumed this link goes to a definition of open-source, not the React github. Suggest putting the link on React (on this line), React library (on the next line), or explicitly add a sentence like:
```suggestion
React is an open-source JavaScript library for building user interfaces. Because it is open-source, we can view its source code in the [React GitHub repository](https://github.com/facebook/react).
```

> The React library will give users the ability to create React _components_.

Reviewer: don't need future tense here
provide a little reason about why we should care about creating these components
```suggestion
The React library gives users the ability to manage UI complexity by creating React _components_.
```

**Components** are rendered UI elements. Components in React may contain:

- Styles
- Complex logic
- Data
- State
- Event-handling

Components are intended to be flexible and reusable, too! A single web page can render many components inside it.

> 

Reviewer: suggest a callout to stress that component is an overloaded term, and _this_ is the meaning we are using for now.
```suggestion

### !callout-info

## Component Is an Over-Used Programming Term

We run into the term _component_ a lot when programming. It doesn't always mean exactly the same thing in every situation. When we are discussing React, we should focus on components as being independent, reusable pieces of UI, regardless of what component might mean in other contexts.

### !end-callout

```

Author: THANK YOU!

## Who is using React?

React launched in 2013, and grew in popularity beginning in 2016.

As of June 2021, the following webapps use React:

- Netflix
- Khan Academy
- Asana
- Facebook
- Airbnb
- Reddit
- UberEats
- PayPal
- Twitter
- Instacart
- OkCupid

## Resources

> Like any new technology, it's great to some resources handy.

Reviewer: ```suggestion
Like any new technology, it's great to have some resources handy.
```
